File: resources/testing/browser_client.py
"""Playwright-driven browser layer on top of the headless trilium_client.py API.

Some things TAM does (fetching a manifest, running `require()`-bundled frontend JSX
widgets, dispatching UI commands) only happen in frontend-env code, which
/api/script/exec can't execute (it only runs backend-env bundles — see
exec_script()'s docstring). To exercise those, drive a real Chromium instance
against the running trilium-server instead.

Usage:
    from browser_client import launch
    with launch() as page:
        page.goto_note(some_note_id)
        page.enable_render_note()  # dismiss the "untrusted render note" warning once
"""
import os
import time
from contextlib import contextmanager

from playwright.sync_api import sync_playwright

PORT = int(os.environ.get("TRILIUM_TESTING_PORT", "8090"))
BASE_URL = f"http://127.0.0.1:{PORT}"

# PLAYWRIGHT_BROWSERS_PATH / PLAYWRIGHT_SKIP_VALIDATE_HOST_REQUIREMENTS are set by
# shell.nix's shellHook (pointing at pkgs.playwright-driver.browsers, matched to the
# pinned `playwright` Python package's expected browser revision) so `playwright
# install` — which tries to download into $HOME/.cache and fails offline/in a
# sandbox — is never needed inside `nix develop`/`nix-shell`.


class TriliumPage:
    """Thin wrapper around a Playwright Page pre-aimed at this trilium-server."""

    def __init__(self, page):
        self.page = page

    def goto_note(self, note_id, wait_seconds=3):
        self.page.goto(f"{BASE_URL}/#root/{note_id}", wait_until="networkidle")
        time.sleep(wait_seconds)
        return self

    def enable_render_note(self, wait_seconds=3):
        """Dismiss the "this render note comes from an external source" trust
        warning Trilium shows the first time a render-type note is opened.
        No-op if the note is already trusted (or isn't a render note)."""
        btns = self.page.get_by_role("button", name="Enable render note")
        if btns.count() > 0:
            btns.first.click(force=True)
            time.sleep(wait_seconds)
        return self

    def __getattr__(self, name):
        return getattr(self.page, name)


@contextmanager
def launch(headless=True):
    """Yields a TriliumPage against a fresh Chromium instance. Closes the
    browser on exit (including on exception)."""
    with sync_playwright() as p:
        browser = p.chromium.launch(args=["--no-sandbox"], headless=headless)
        page = browser.new_page()
        try:
            yield TriliumPage(page)
        finally:
            browser.close()

File: resources/testing/browser_client.js
#!/usr/bin/env node
"use strict";
/* Playwright-driven browser layer on top of the headless trilium_client.js API.

Some things TAM does (fetching a manifest, running `require()`-bundled frontend JSX
widgets, dispatching UI commands) only happen in frontend-env code, which
/api/script/exec can't execute (it only runs backend-env bundles -- see
execScript()'s docstring). To exercise those, drive a real Chromium instance
against the running trilium-server instead.

Usage:
    const { withPage } = require("./browser_client");
    await withPage(async (page) => {
        await page.gotoNote(someNoteId);
        await page.enableRenderNote();  // dismiss the "untrusted render note" warning once
    });
*/

const { chromium } = require("playwright");

const PORT = parseInt(process.env.TRILIUM_TESTING_PORT || "8090", 10);
const BASE_URL = `http://127.0.0.1:${PORT}`;

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

// PLAYWRIGHT_BROWSERS_PATH / PLAYWRIGHT_SKIP_VALIDATE_HOST_REQUIREMENTS are set by
// shell.nix's shellHook (pointing at pkgs.playwright-driver.browsers, matched to the
// pinned `playwright` package's expected browser revision) so `playwright install`
// -- which tries to download into $HOME/.cache and fails offline/in a sandbox -- is
// never needed inside `nix develop`/`nix-shell`.

class TriliumPage {
    // Thin wrapper around a Playwright Page pre-aimed at this trilium-server.
    // Unknown methods/properties fall through to the underlying Page via Proxy.
    constructor(page) {
        this.page = page;
        return new Proxy(this, {
            get(target, prop, receiver) {
                if (prop in target) return Reflect.get(target, prop, receiver);
                const val = target.page[prop];
                return typeof val === "function" ? val.bind(target.page) : val;
            },
        });
    }

    async gotoNote(noteId, waitSeconds = 3) {
        await this.page.goto(`${BASE_URL}/#root/${noteId}`, { waitUntil: "networkidle" });
        await sleep(waitSeconds * 1000);
        return this;
    }

    async enableRenderNote(waitSeconds = 3) {
        // Dismiss the "this render note comes from an external source" trust
        // warning Trilium shows the first time a render-type note is opened.
        // No-op if the note is already trusted (or isn't a render note).
        const btns = this.page.getByRole("button", { name: "Enable render note" });
        if (await btns.count() > 0) {
            await btns.first().click({ force: true });
            await sleep(waitSeconds * 1000);
        }
        return this;
    }
}

async function withPage(fn, { headless = true } = {}) {
    // Runs fn(page) against a fresh Chromium instance. Closes the browser on
    // exit (including on exception).
    const browser = await chromium.launch({ args: ["--no-sandbox"], headless });
    try {
        const page = await browser.newPage();
        return await fn(new TriliumPage(page));
    } finally {
        await browser.close();
    }
}

module.exports = { withPage, TriliumPage, BASE_URL, PORT };
